internal/buffer/backup.go

```diff
@@ -34,6 +30,22 @@ Options: [r]ecover, [i]gnore, [a]bort: `
 
 var backupRequestChan chan *Buffer
 
+type BackupWriter struct {
```

Comment:
Backing up the backup? Do you think it's a good idea?

Reply:
Well, maybe it is...

Reply:
Or maybe not... I suggest to do a thought experiment on what exactly happens to the buffer and its backup over its lifetime (first background backup, second background backup, file save, first background backup after file save, and so on), to figure out if your PR's behavior is the desired one, and what should be the desired behavior.

Reply:
1. Fixed regression with "absolute" paths received via save.
2. Try to prevent async backups in case the recently performed save was successful.

It can be discussed, if the backup of the backup is really needed. At least it's a further stage in the moment `permbackup` is set, but the backup can't be written...so we wouldn't truncate the last valid backup.

Reply:
Anything to add? :thinking:

Reply:
Had no time yet, sorry.

Since you're asking, just some preliminary high-level thoughts:
- My feeling is that backup for a backup is too much. And, actually, backup for a serialized buffer file in `~/.config/micro/buffers` too. You are right it's a good idea to avoid irrecoverably corrupting those files as well. But those files (backups and serialized buffers) are for micro's own internal use, the user is not supposed to do anything with them directly. So it seems better if we just `os.Rename()` them instead (i.e. do what we originally intended to do for other files). That is simpler, more reliable (since it's atomic) and transparent to the user. We don't need to care if e.g. a file in `~/.config/micro/backups` is suddenly a symlink and we silently replace it by `os.Rename()`, and so on. The user is not supposed to manually mess with whatever is inside `~/.config/micro/backups` and `~/.config/micro/buffers` directories.
- Your PR still doesn't implement *applying* the backups for broken `settings.json` and `bindings.json`, am I right? And without that, those backups are useless, right? (Unless the user enables `permbackup` and recovers them manually, but that's an obscure corner case.) I imagine we should check and apply backups for `settings.json` and `bindings.json` every time when micro starts (before it opens any files, etc): if there is a backup, ask the user it we should recover the settings/bindings or not, like we do for buffers. (We probably need to do that very early, before we initialize settings and bindings.)
- Actually I'm not sure it's a good idea to provide `permbackup` for `settings.json` and `bindings.json` backups.
- I'm still not sure it's a good idea to mix backups for `settings.json` and `bindings.json` as config files with backups for the same files as buffers edited by the user (aka "reuse" the same backup files for both purposes).

Reply:
> My feeling is that backup for a backup is too much. And, actually, backup for a serialized buffer file in `~/.config/micro/buffers` too.
> [...]
> So it seems better if we just `os.Rename()` them instead (i.e. do what we originally intended to do for other files).

But even then we've to use some kind of intermediate file with a new dedicated name. So why not directly use the same new generic mechanism we already have? I give you right, that keeping a backup of the backup is senseless, which was already the reason why I returned a hard-coded `false` via `KeepBackup()` in the `BackupWriter` (it's removed afterwards anyway). The same should be done for the `SerializedWriter`, where it currently isn't...it shall not keep a backup of that; check!
Furthermore the temporary backup of the backup is only created for the asynchronous backup task and not for the already intermediate step of the target file. Maybe something was misunderstood meanwhile.

> Actually I'm not sure it's a good idea to provide `permbackup` for `settings.json` and `bindings.json` backups.

We can set it to a hard-coded `false` for these two types of files too and revert the `permbackup` conversion into a global option. Maybe this is the better choice here.

> I'm still not sure it's a good idea to mix backups for `settings.json` and `bindings.json` [...]

Alternatively we could point to the same directory of these files. In case we agree the commit https://github.com/zyedidia/micro/pull/3273/commits/efbab5a28e38061f65d7de25a1e204967fd63084 is then superfluous.

I will see what and when I can do the other points.

Reply:
> So why not directly use the same new generic mechanism we already have?

Because this mechanism is not atomic and thus not transparent to the user. We may end up corrupting the original file when overwriting if with the backup. What do we do then? We tell the user "A backup was detected for this file. This likely means..." and expect the user to choose one of [r]ecover, [i]gnore, [a]bort. Do we want to do that for backup files and serialized buffers?

Or we may not ask the user and just recover the backup file automatically. But what if it was the other way around: the backup file was not written successfully (but was not removed after that, because micro crashed) and as a result, contains incorrect data? We would silently overwrite a correct backup with an incorrect one?

Reply:
Ok, then we end up in having `SafeWrite()` and the `FileWriter` interface for the normal `buffer` save path only (at least so far), since all other file types (backups, serialized buffers, bindings and settings) can be stored much easier without `SafeWrite` via a save to a temporary file in the same directory followed by a rename.

But how to handle a possible failed rename in these scenarios?

- drop or keep the new temporary file and return the error?
- try to write into the target file directly and risk a corruption of that?

Reply:
> since all other file types (backups, serialized buffers, bindings and settings) can be stored much easier without `SafeWrite` via a save to a temporary file in the same directory followed by a rename.

No, bindings and settings are a different case: they are edited and managed by the user, and the user may want it to be a symlink, and so on. (See #3016, for example.)

That said, it is a rather good idea to *first* address safe saving for normal buffers only, and *after that* generalize it to settings and bindings. It seems that in this PR you jumped straight into an attempt to solve both problems at once, which is difficult.

> But how to handle a possible failed rename in these scenarios?

Dropping the new temporary file and returning the error seems to be the most reasonable behavior.

Reply:
The renaming is now used for the backups and serialized buffers only.
The logic for the recovery is still not done yet.

Reply:
> ~Maybe adding a SHA256 hash at the end of a backup or serialized buffer file could become problematic, since the file name already consists of the absolute path and thus the whole name can be too long. I assume a shorter hash is needed to represent a unique and reproducible temporary extension.~
> `base64` should be feasible enough as reproducible extension.

Why do we need anything of that at all? Can't we just use the same extension e.g. `.micro-backup` (i.e. something that is unlikely to be used for anything else, unlike `.tmp` and so on) for all such files?

If the file opened by the user is this temporary file itself, e.g. `settings.json.micro-backup`, it shouldn't be a problem, since micro, in turn, will look for `settings.json.micro-backup.micro-backup`.